store/mongo/datastore_mongo.go

```diff
@@ -360,38 +422,110 @@ func (db *DataStoreMongo) GetSessionRecording(ctx context.Context,
 	if err != nil {
 		return err
 	}
+	defer recordingsCursor.Close(ctx)
 
-	output := make([]byte, recordingReadBufferSize)
-	var buffer bytes.Buffer
-	for c.Next(ctx) {
-		var r model.Recording
-		err = c.Decode(&r)
-		if err != nil {
-			return err
-		}
-
-		buffer.Reset()
-		buffer.Write(r.Recording)
-		gzipReader, e := gzip.NewReader(&buffer)
-		if e != nil {
-			err = e
-		}
-
-		for {
-			n, e := gzipReader.Read(output)
-			if n == 0 || e != nil {
-				gzipReader.Close()
+	controlCursor, err := collControl.Find(ctx,
+		bson.M{
+			dbFieldSessionID: sessionID,
+		},
+		findOptions,
+	)
+	if err != nil {
+		return err
+	}
+	defer controlCursor.Close(ctx)
+
+	controlReader := NewControlMessageReader(ctx, controlCursor)
+	recordingReader := NewRecordingReader(ctx, recordingsCursor)
+
+	recordingBuffer := make([]byte, recordingReadBufferSize)
+	recordingBytesSent := 0
+	for {
+		control := controlReader.Pop()
+		if control == nil {
+			l.Debug("WriteSessionRecords: no more control " +
+				"messages, flushing the recording upstream.")
+			//no more control messages, we send the whole recording
+			var err error = nil
+			var n int
+			for err != io.EOF {
+				n, err = recordingReader.Read(recordingBuffer)
+				if n > 0 {
+					//nolint:errcheck
+					sendRecordingMessage(recordingBuffer[:n], sessionID, w)
+					recordingBytesSent += n
+				}
+			}
+			break
+		} else {
+			if recordingBytesSent > control.Offset {
+				//this should never happen, we missed
+				//the control message, no idea what to do
+				l.Errorf("WriteSessionRecords: recordingBytesSent > control.Offset")
 				break
 			}
-			_, e = w.Write(output[:n])
-			if e != nil {
-				err = e
+
+			bytesUntilControlMessage := control.Offset - recordingBytesSent
+			l.Debugf("(1) WriteSessionRecords: control.Offset:%d"+
+				" recordingBytesSent:%d "+
+				" bytesUntilControlMessage:%d (recordingBuffer.len=%d) "+
+				"reading up to %d bytes of recording and sending.",
+				control.Offset,
+				recordingBytesSent,
+				bytesUntilControlMessage,
+				len(recordingBuffer),
+				control.Offset-recordingBytesSent)
+			//it is possible that the recording is larger than one recordingBuffer,
+			//we need to send until we have the control.Offset in the buffer
+			for bytesUntilControlMessage > len(recordingBuffer) {
+				n, e := recordingReader.Read(recordingBuffer)
+				if n > 0 {
+					//error handling: see below
+					//nolint:errcheck
```

Comment:
You SHOULD check these errors.

Reply:
fixed. breaking and returning on each one of them.